c/tskit/trees.c

```diff
@@ -5673,111 +5735,57 @@ tsk_tree_position_in_interval(const tsk_tree_t *self, double x)
- * in linear time and requires no additional memory.
- *
- * During design, the following alternatives were tested
- * (in a combination of rust + C):
- * 1. Indexing edge insertion/removal locations by tree.
- *    The indexing can be done in O(n) time, giving O(1)
- *    access to the first edge in a tree. We can then add
- *    edges to the tree in O(e) time, where e is the number
- *    of edges. This apparoach requires O(n) additional memory
- *    and is only marginally faster than the implementation below.
- * 2. Building an interval tree mapping edge id -> span.
- *    This approach adds a lot of complexity and wasn't any faster
- *    than the indexing described above.
- */
 static int
 tsk_tree_seek_from_null(tsk_tree_t *self, double x, tsk_flags_t TSK_UNUSED(options))
 {
     int ret = 0;
-    tsk_size_t edge;
-    tsk_id_t p, c, e, j, k, tree_index;
-    const double L = tsk_treeseq_get_sequence_length(self->tree_sequence);
-    const tsk_treeseq_t *treeseq = self->tree_sequence;
-    const tsk_table_collection_t *tables = treeseq->tables;
+    tsk_table_collection_t *tables = self->tree_sequence->tables;
     const tsk_id_t *restrict edge_parent = tables->edges.parent;
     const tsk_id_t *restrict edge_child = tables->edges.child;
-    const tsk_size_t num_edges = tables->edges.num_rows;
-    const tsk_size_t num_trees = self->tree_sequence->num_trees;
     const double *restrict edge_left = tables->edges.left;
     const double *restrict edge_right = tables->edges.right;
-    const double *restrict breakpoints = treeseq->breakpoints;
-    const tsk_id_t *restrict insertion = tables->indexes.edge_insertion_order;
-    const tsk_id_t *restrict removal = tables->indexes.edge_removal_order;
-
-    // NOTE: it may be better to get the
-    // index first and then ask if we are
-    // searching in the first or last 1/2
-    // of trees.
-    j = -1;
+    double interval_left = self->interval.left;
+    double interval_right = self->interval.right;
+    const double *restrict breakpoints = self->tree_sequence->breakpoints;
+    const tsk_size_t num_trees = self->tree_sequence->num_trees;
+    const double L = tsk_treeseq_get_sequence_length(self->tree_sequence);
+    tsk_id_t j, e, index;
+    tsk_tree_position_t tree_pos;
+
+    index = (tsk_id_t) tsk_search_sorted(breakpoints, num_trees + 1, x);
+    if (breakpoints[index] > x) {
+        index--;
+    }
+
     if (x <= L / 2.0) {
-        for (edge = 0; edge < num_edges; edge++) {
-            e = insertion[edge];
-            if (edge_left[e] > x) {
-                j = (tsk_id_t) edge;
-                break;
-            }
-            if (x >= edge_left[e] && x < edge_right[e]) {
-                p = edge_parent[e];
-                c = edge_child[e];
-                tsk_tree_insert_edge(self, p, c, e);
+        ret = tsk_tree_position_seek_forward(&self->tree_pos, index);
+        if (ret != 0) {
+            goto out;
+        }
```

Comment:
It took me a couple of minutes to realise that because we're seeking from the null tree there are no edges to remove. It's worth adding a comment to that effect here I think.